### PR TITLE
Add hyperparameters support into Decision Forest training algorithm on CPU

### PR DESCRIPTION
**Pull request by Vika-F**
## Description

Two hyperparameters were added to Decision Forest regression training algorithm on CPU:

- minPartCoefficient: Multiplier that defines the minimum work size for a thread to be assigned a separate task. For example, the value 4 means that a thread will be assigned a separate task that processes at least 4 topmost levels of a tree.
- minSizeCoefficient: Another multiplier that defines the minimum work size for a thread to be assigned a separate task. For example, the value 24000 means that a thread will be assigned a separate task that processes a 8-level tree built on at least 1000 observations and 10 features.

The resulting minimum work size is defined as the maximum of two values computed using those two constants.

Three hyperparameters were added to Decision Forest classification training algorithm on CPU:

- The two values are similar to the regression algorithm, and
- smallNClassesThreshold: The threshold value to determine if the number of classes is small enough to switch to the implementation that is optimized for small number of classes.

The APIs for hyperparameters setting and getting was added to oneDAL part.
The mechanism that propagates the hyperparameters values to DAAL computational kernels was implemented.
The test that varies the values of hyperparameters and checks the training results was added.

Additionally, dead code the was used in Java build previously was removed from `makefile.lst`.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.

  The documentation was not updated because we do not document preview features located in `detail` namespace.

- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

Not applicable.

### CODE REVIEW

**Comment by Vika-F** [comment]
Pre-commit: http://intel-ci.intel.com/eff28fba-f8d7-f1ac-85d2-a4bf010d0e2d

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i [hunk 1]
```diff
@@ -78,9 +78,17 @@ public:
     typedef SplitData<algorithmFPType, ImpurityData> TSplitData;
 
 public:
-    UnorderedRespHelperBest(const dtrees::internal::IndexedFeatures * indexedFeatures, size_t nClasses)
+    UnorderedRespHelperBest(const dtrees::internal::IndexedFeatures * indexedFeatures, size_t nClasses,
+                            const classification::training::internal::Hyperparameter * hp = nullptr)
         : super(indexedFeatures), _nClasses(nClasses), _histLeft(nClasses), _impLeft(nClasses), _impRight(nClasses)
-    {}
+    {
+        DAAL_INT64 nClassesThreshold = 8l;
```
Comment (david-cortes-intel) [inline cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i:85]: Could this one be obtained by default-initializing the struct that has the defaults in the declaration? Same for the 'propose' functions.

File: cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter.cpp [hunk 2]
```diff
@@ -30,6 +30,70 @@ namespace algorithms
...
+{
+namespace training
+{
+namespace internal
+{
+Hyperparameter::Hyperparameter() : algorithms::Hyperparameter(hyperparameterIdCount, doubleHyperparameterIdCount) {}
+
+services::Status Hyperparameter::set(HyperparameterId id, DAAL_INT64 value)
+{
+    return this->algorithms::Hyperparameter::set(uint32_t(id), value);
```
Comment (david-cortes-intel) [inline cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter.cpp:43]: Don't the find/set methods use `uint64_t`?

File: cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter_impl.h [hunk 3]
```diff
@@ -33,11 +33,182 @@ namespace algorithms
...
+
+/**
+ * Available identifiers of integer hyperparameters of the decision forest training algorithm
+ */
+enum HyperparameterId
+{
+    smallNClassesThreshold = 0,
+    minPartCoefficient,
+    minSizeCoefficient,
+    hyperparameterIdCount = minSizeCoefficient + 1
```
Comment (david-cortes-intel) [inline cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter_impl.h:51]: I think better to explicitly enumerate these. As far as I'm aware, it's not required for the compiler to produce an incremental sequence for internal mapping, even if it's currently doing so.

File: cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter_impl.h [hunk 4]
```diff
@@ -33,11 +33,182 @@ namespace algorithms
...
+    {
+        DAAL_INT64 smallNClassesValue = 0l;
+        DAAL_INT64 minPartCoeffValue  = 0l;
+        DAAL_INT64 minSizeCoeffValue  = 0l;
+        s |= find(smallNClassesThreshold, smallNClassesValue);
+        s |= find(minPartCoefficient, minPartCoeffValue);
+        s |= find(minSizeCoefficient, minSizeCoeffValue);
+        if (!s) return;
+
+        if (smallNClassesValue < 1 || 8l < smallNClassesValue || minPartCoeffValue < 1 || minSizeCoeffValue < 1)
```
Comment (david-cortes-intel) [inline cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter_impl.h:111]: Does '8' come from some constrain in the code logic? If so, could be helpful to leave a comment here with an explanation and a warning not to alter it.

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i (hunk 1 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i]: The struct is defined in oneDAL, and DAAL knows nothing about it.

This initialization is only needed here for backward compatibility reasons. Because if the binary was compiled with the older version of oneDAL, `hp` can be `nullptr`; so the default initialization is required.

It would be possible to remove this initialization and NULL pointer check in the next gold release.
For now I've defined the default value on DAAL side:
https://github.com/uxlfoundation/oneDAL/pull/3063/files#diff-453a712bfb029f1f4c7ca37eeb488001f9291470d2f74199d700ac74ab8e27c6R45

File: cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter.cpp (hunk 2 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter.cpp:43]: No, in DAAL they use uint32_t:
https://github.com/uxlfoundation/oneDAL/blob/main/cpp/daal/src/algorithms/algorithm_hyperparameter.cpp#L84

File: cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter_impl.h (hunk 3 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter_impl.h]: Done.

File: cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter_impl.h (hunk 4 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter_impl.h]: Added the comment here:
https://github.com/uxlfoundation/oneDAL/pull/3063/files#diff-453a712bfb029f1f4c7ca37eeb488001f9291470d2f74199d700ac74ab8e27c6R45

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i [hunk 5]
```diff
@@ -634,8 +638,25 @@ protected:
 
     void setupHostApp()
     {
-        const size_t minPart = 4 * _helper.size();        //corresponds to the 4 topmost levels
-        const size_t minSize = 24000 / _nFeaturesPerNode; //at least that many, corresponds to the tree 1000 obs/10 features/8 levels
+        /// const size_t minPart = 4 * _helper.size();         //corresponds to the 4 topmost levels
```
Comment (david-cortes-intel) [inline cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i:641]: Were these comments meant to go into the latter lines?

File: cpp/oneapi/dal/algo/decision_forest/detail/train_ops.hpp [hunk 6]
```diff
@@ -75,14 +84,46 @@ struct train_ops {
...
+    template <typename ParameterType>
+    void check_parameters_ranges(const ParameterType& params, const input_t& input) const {
+        if constexpr (std::is_same_v<task_t, task::classification>) {
+            ONEDAL_ASSERT(params.get_small_classes_threshold() > 0);
+            ONEDAL_ASSERT(params.get_small_classes_threshold() <= 8);
+        }
+        ONEDAL_ASSERT(params.get_min_part_coefficient() > 0);
+        ONEDAL_ASSERT(params.get_min_part_coefficient() <= 256);
+        ONEDAL_ASSERT(params.get_min_size_coefficient() > 0);
+        ONEDAL_ASSERT(params.get_min_size_coefficient() <= 0x100000l);
```
Comment (david-cortes-intel) [inline cpp/oneapi/dal/algo/decision_forest/detail/train_ops.hpp:97]: Shouldn't this one also get checked in `cpp/daal/src/algorithms/dtrees/forest/df_hyperparameter_impl.h`?

File: cpp/oneapi/dal/algo/decision_forest/detail/train_ops.hpp [hunk 7]
```diff
@@ -75,14 +84,46 @@ struct train_ops {
                               const input_t& input,
                               const result_t& result) const {}
 
+    /// Check that the hyperparameters of the algorithm belong to the expected ranges
+    template <typename ParameterType>
+    void check_parameters_ranges(const ParameterType& params, const input_t& input) const {
+        if constexpr (std::is_same_v<task_t, task::classification>) {
+            ONEDAL_ASSERT(params.get_small_classes_threshold() > 0);
+            ONEDAL_ASSERT(params.get_small_classes_threshold() <= 8);
```
Comment (david-cortes-intel) [inline cpp/oneapi/dal/algo/decision_forest/detail/train_ops.hpp:92]: Since this one is onedal instead of daal, is it possible to take the value from the daal headers?

**Comment by david-cortes-intel** [comment]
For some reason I'm unable to access the error logs where it failed, but it looks like the failures were in algorithms unrelated to decision trees.

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i (hunk 5 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i]: Moved the comments to where they belong.

File: cpp/oneapi/dal/algo/decision_forest/detail/train_ops.hpp (hunk 6 above)
Comment (Vika-F) [reply cpp/oneapi/dal/algo/decision_forest/detail/train_ops.hpp]: Yes, but here the check are compile-time and enabled only in debug build. And in DAAL the checks are runtime and enabled always. I prefer to have both checks implemented.

File: cpp/oneapi/dal/algo/decision_forest/detail/train_ops.hpp (hunk 7 above)
Comment (Vika-F) [reply cpp/oneapi/dal/algo/decision_forest/detail/train_ops.hpp]: Done.

**Review by david-cortes-intel — APPROVED** [review]

**Comment by david-cortes-intel** [comment]
Failure from the CI:
```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
In file included from external/onedal_release/include/oneapi/dal/algo/decision_forest/train.hpp:19,
                 from external/onedal_release/include/oneapi/dal/algo/decision_forest.hpp:20,
                 from examples/oneapi/cpp/source/decision_forest/df_reg_dense_batch.cpp:18:
external/onedal_release/include/oneapi/dal/algo/decision_forest/detail/train_ops.hpp:19:10: fatal error: daal/src/algorithms/dtrees/forest/df_hyperparameter_impl.h: No such file or directory
   19 | #include <daal/src/algorithms/dtrees/forest/df_hyperparameter_impl.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: cpp/oneapi/dal/algo/decision_forest/train_types.hpp [hunk 8]
```diff
@@ -58,6 +58,9 @@ class ONEDAL_EXPORT train_parameters<task::regression> : public dal::detail::sys
         return *this;
     }
 
+    /// Check that the hyperparameters of the algorithm belong to the expected ranges
+    void check() const;
```
Comment (david-cortes-intel) [inline cpp/oneapi/dal/algo/decision_forest/train_types.hpp:62]: How "check_parameters_are_valid" or something along those lines?

Comment (Vika-F) [reply cpp/oneapi/dal/algo/decision_forest/train_types.hpp:62]: I think the proposed name is ambiguous. For me "check_ranges" or "check_correctness" or "check_validity" would be better.

Comment (david-cortes-intel) [reply cpp/oneapi/dal/algo/decision_forest/train_types.hpp:62]: Sure, those sounds more descriptive than just 'check'.

**Review by david-cortes-intel — APPROVED** [review]
